rust/otap-dataflow/crates/otap/src/pdata.rs

```diff
@@ -129,85 +137,272 @@ impl OtlpProtoBytes {
 /// Container for the various representations of the telemetry data
 #[derive(Clone, Debug)]
 #[allow(clippy::large_enum_variant)]
-pub enum OtapPdata {
+pub enum OtapPayload {
     /// data is serialized as a protobuf service message for one of the OTLP GRPC services
     OtlpBytes(OtlpProtoBytes),
 
     /// data is contained in `OtapBatch` which contains Arrow `RecordBatches` for OTAP payload type
     OtapArrowRecords(OtapArrowRecords),
 }
 
+/// Context + container for telemetry data
+#[derive(Clone, Debug)]
+pub struct OtapPdata {
+    context: Context,
+    payload: OtapPayload,
+}
+
+/* -------- Signal type -------- */
+
 impl OtapPdata {
+    /// Construct new OtapData with payload using default context.
+    #[must_use]
+    pub fn new_default(payload: OtapPayload) -> Self {
+        Self {
+            context: Context::default(),
+            payload,
+        }
+    }
+
+    /// Construct new OtapData with context and payload
+    #[must_use]
+    pub fn new(context: Context, payload: OtapPayload) -> Self {
+        Self { context, payload }
+    }
+
+    /// Returns the type of signal represented by this `OtapPdata` instance.
+    #[must_use]
+    pub fn signal_type(&self) -> SignalType {
+        self.payload.signal_type()
+    }
+
+    /// True if the payload is empty. By definition, we can skip sending an
+    /// empty request.
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.payload.is_empty()
+    }
+
+    /// Returns the payload from this request, consuming it.  This is
+    /// only considered useful in testing.  Use split() to split an
+    /// OtapPdata into (Context, OtapPayload).
+    #[must_use]
+    #[cfg(test)]
+    pub fn payload(self) -> OtapPayload {
+        self.payload
+    }
+
+    /// Splits the context and payload from this request, consuming it.
+    #[must_use]
+    pub fn split(self) -> (Context, OtapPayload) {
+        (self.context, self.payload)
+    }
+
+    /// Clones the payload from this request.
+    #[must_use]
+    pub fn clone_payload(&self) -> OtapPayload {
+        self.payload.clone()
+    }
```

Comment:
For me, these methods are temporary and I don't think we should ultimately expose them at all for the nodes. I'll create a GitHub issue to describe what I have in mind at the pdata level to avoid this. But for now, I don't want to block you.